v2/internal/pkg/batch/concurrent_chan_worker.go

```diff
@@ -0,0 +1,383 @@
+							}
+							results <- result
+							break loop
+						}
+
+					}
+				}
+			}(cancelCtx, semaphore, results, sp)
+		}
+		wg.Wait()
+	}()
+
+	overallProgress := newOverallProgress(p, total)
+
+	go runOverallProgress(overallProgress, cancelCtx, progressCh)
+
+	completed := 0
+	for completed < len(collectorSchema.AllImages) {
+		res := <-results
+		err := res.err
+		if err == nil {
+			copiedImages.AllImages = append(copiedImages.AllImages, res.img)
+			incrementTotals(res.imgType, &copiedImages)
+
+			//OCPBUGS-44448
+			if opts.IsDelete() {
+				m.Lock()
+				alreadyDeleted[res.img.Origin] = struct{}{}
+				m.Unlock()
+			}
+
+		} else {
+			m.Lock()
+			errArray = append(errArray, *err)
+			m.Unlock()
+
+			if res.imgType.IsRelease() {
+				cancel()
+				break
+			}
+		}
+
+		completed++
+		progressCh <- 1
+	}
+	close(progressCh)
+
+	p.Wait()
+
+	logResults(o.Log, opts.Function, &copiedImages, &collectorSchema)
+
+	if len(errArray) > 0 {
+		filename, err := saveErrors(o.Log, o.LogsDir, errArray)
+		if err != nil {
+			return copiedImages, NewSafeError(errMsgHeader+" - unable to log these errors in %s/%s: %s", workerPrefix, o.LogsDir, filename, err.Error())
+		} else {
+			msg := fmt.Sprintf(errMsg, workerPrefix, o.LogsDir, filename)
+			return copiedImages, NewSafeError(msg)
+		}
+	}
+	endTime := time.Now()
+	execTime := endTime.Sub(startTime)
+	o.Log.Debug("concurrent channel worker time     : %v", execTime)
+	return collectorSchema, nil
+}
+
+func hostNamespace(input string) string {
+	parsedURL, err := url.Parse(input)
+	if err != nil {
+		return ""
+	}
+
+	host := parsedURL.Host
+	pathSegments := strings.Split(strings.Trim(parsedURL.Path, "/"), "/")
+
+	if len(pathSegments) > 1 {
+		namespacePath := strings.Join(pathSegments[:len(pathSegments)-1], "/")
+		hostAndNamespace := path.Join(host, namespacePath) + "/"
+		return hostAndNamespace
+	} else if len(pathSegments) == 1 {
+		return path.Join(host, pathSegments[0]) + "/"
+	} else {
+		return host
+	}
+}
+
+func logResults(log clog.PluggableLoggerInterface, copyMode string, copiedImages, collectorSchema *v2alpha1.CollectorSchema) {
+	log.Info("=== Results ===")
+
+	var copyModeMsg string
+	if copyMode == string(mirror.CopyMode) {
+		copyModeMsg = "mirrored"
+	} else {
+		copyModeMsg = "deleted"
+	}
+
+	logResult(log, copyModeMsg, "release", copiedImages.TotalReleaseImages, collectorSchema.TotalReleaseImages)
+	logResult(log, copyModeMsg, "operator", copiedImages.TotalOperatorImages, collectorSchema.TotalOperatorImages)
+	logResult(log, copyModeMsg, "additional", copiedImages.TotalAdditionalImages, collectorSchema.TotalAdditionalImages)
+	logResult(log, copyModeMsg, "helm", copiedImages.TotalHelmImages, collectorSchema.TotalHelmImages)
+}
+
+func logResult(log clog.PluggableLoggerInterface, copyMode, imageType string, copied, total int) {
+	if total != 0 {
+		if copied == total {
+			log.Info("✅ %d / %d %s images %s successfully", copied, total, imageType, copyMode)
```

Comment:
@aguidirh -nit: just for consistency  could we use the same arrow and cross thats used in the spinner ?

Reply:
I like the idea @lmzuccarelli, so I just pushed (in this PR) a new package where we can keep all the emojis and reuse them from there.

In this way it is easier to change an emoji when needed and also to find/add/remove them and needed.

Reply:
:100: